torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -837,7 +823,12 @@ def _post_backward_hook(
                     flat_param._saved_grad_shard = new_sharded_grad
                 grad_to_offload = flat_param._saved_grad_shard
             else:
-                state._comm_hook(state._comm_hook_state, flat_param.grad)
+                if state._comm_hook is None:  # default path
```

Comment:
is this the  no_shard case? can we unify these paths in the future?

Reply:
I think this is already pretty unified, and further unification may be challenging. All-reduce operates in-place without need for allocating the new sharded gradient tensor or for padding the unsharded gradient tensor. The gradient accumulation happens directly.